### PR TITLE
fix: Optimization of dialog validation

### LINKED ISSUE #4203: Optimize unnecessary extra validations (eg arrays of items that cause validation on every children)

### PR DESCRIPTION
**Pull request by LouisEugeneMSFT**
## Description

Most recoil atoms in the Composer state are tracking data through arrays. In the case of dialogs, this was problematic since all dialogs were being validated any time any of them had any change (say a new character in a field). 

This PR addresses this issue by introducing an atom family tracking the individual dialogs. Any creation, update, validation, deletion can now target an individual dialog instead of modifying the entire array of dialogs. To be able to still list all the dialogs, an array of dialogIds is introduced. All the dialogs data can still be accessed, but it is now through a selector rather than an atom.

image

## Task Item

fixes #4203

### CODE REVIEW

File: Composer/packages/client/src/recoilModel/selectors/dialogs.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { DialogInfo } from '@bfc/shared';
+import { selectorFamily } from 'recoil';
+
+import { dialogIdsState, dialogState } from '../atoms';
+
+export const dialogsState = selectorFamily<DialogInfo[], string>({
```
Comment (hatpick) [inline Composer/packages/client/src/recoilModel/selectors/dialogs.ts:9]: the name better to indicate that it's a selectorFamily

**Review by hatpick — APPROVED** [review]

File: Composer/packages/client/src/pages/knowledge-base/QnAPage.tsx [hunk 2]
```diff
@@ -14,7 +14,7 @@ import { navigateTo } from '../../utils/navigation';
 import { TestController } from '../../components/TestController/TestController';
 import { INavTreeItem } from '../../components/NavTree';
 import { Page } from '../../components/Page';
-import { dialogsState, qnaFilesState } from '../../recoilModel/atoms/botState';
```
Comment (GeoffCoxMSFT) [inline Composer/packages/client/src/pages/knowledge-base/QnAPage.tsx:17]: I don't think we should have an index barrel file for the recoilModel.  We found in LUIS how much is slows down build and messes with hot module reload.

File: Composer/packages/client/src/recoilModel/DispatcherWrapper.tsx [hunk 3]
```diff
@@ -10,11 +10,11 @@ import { BotAssets } from '@bfc/shared';
 import { useRecoilValue } from 'recoil';
 import isEmpty from 'lodash/isEmpty';
 
+import { dialogsState } from './selectors';
```
Comment (GeoffCoxMSFT) [inline Composer/packages/client/src/recoilModel/DispatcherWrapper.tsx]: I know you're trying to have minimal change impact, but if it is a selector it should probably be called a selector.

File: Composer/packages/client/src/recoilModel/selectors/dialogs.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  key: 'dialogs',
+  get: (projectId: string) => ({ get }) => {
+    const dialogIds = get(dialogIdsState(projectId));
+
+    return dialogIds.map((dialogId) => {
+      return get(dialogState({ projectId, dialogId }));
+    });
+  },
+  set: (projectId: string) => ({ set }, newDialogs) => {
```
Comment (GeoffCoxMSFT) [inline Composer/packages/client/src/recoilModel/selectors/dialogs.ts:18]: Nit: I'm not a fan of the set on a selector, but I can't claim it won't work.

Comment (GeoffCoxMSFT) [inline Composer/packages/client/src/recoilModel/selectors/dialogs.ts:18]: Isn't newDialogs expected to be an array (according to Typescript)?  Do you need to defend / silently fail if someone passes a non-array?

File: Composer/packages/client/src/recoilModel/selectors/dialogs.ts (hunk 1 above)
Comment (LouisEugeneMSFT) [reply Composer/packages/client/src/recoilModel/selectors/dialogs.ts]: done

**Review by GeoffCoxMSFT — APPROVED** [review]

**Review by GeoffCoxMSFT — APPROVED** [review]

**Pull request merged**